README.rst
=======
JumpSSH
=======

.. image:: https://secure.travis-ci.org/AmadeusITGroup/JumpSSH.svg?branch=master
    :target: http://travis-ci.org/AmadeusITGroup/JumpSSH

.. image:: https://coveralls.io/repos/AmadeusITGroup/JumpSSH/badge.svg?branch=master
    :target: https://coveralls.io/r/AmadeusITGroup/JumpSSH?branch=master

.. image:: https://badge.fury.io/py/jumpssh.svg
    :target: https://badge.fury.io/py/jumpssh

.. image:: https://readthedocs.org/projects/jumpssh/badge?version=latest
    :target: https://jumpssh.readthedocs.io?badge=latest

.. image:: https://sonarcloud.io/api/project_badges/measure?project=amadeusitgroup_jumpssh&metric=bugs
    :target: https://sonarcloud.io/dashboard?id=amadeusitgroup_jumpssh

.. image:: https://sonarcloud.io/api/project_badges/measure?project=amadeusitgroup_jumpssh&metric=vulnerabilities
    :target: https://sonarcloud.io/dashboard?id=amadeusitgroup_jumpssh


:JumpSSH:          Python module to run commands on remote servers
:Copyright:        Copyright (c) 2017 Amadeus sas
:License:          `MIT <https://github.com/AmadeusITGroup/JumpSSH/blob/master/LICENSE>`_
:Documentation:    https://jumpssh.readthedocs.io
:Development:      https://github.com/AmadeusITGroup/JumpSSH

What
----
`JumpSSH` is a module for Python 2.7+/3.5+ that can be used to run commands on remote servers through a gateway.

It is based on `paramiko library <http://www.paramiko.org>`_.
It provides the ability to execute commands on hosts that are not directly accessible but only through one or
more servers.
Script does not need to be uploaded on a remote server and can be run locally.

Several authentication methods are supported (password, ssh key).

Commands can be run through several jump servers before reaching the remote server.
No need to establish a session for each command, a single ssh session can run as many command as you want,
including parallel queries, and you will get result for each command independently.

So, why another python library to setup remote server through ssh ? Here is a quick comparison with the most known existing python libraries
 - Paramiko: provide very good implementation of SSHv2 protocol in python but with a low level api a bit complex
 - Ansible: require more configuration and understanding to start.
   Moreover, support of bastion host is done with modification of local ssh config to use ProxyCommand, and this is
   needed for each bastion host.
 - Fabric: use of jump server is much easier than Ansible thanks to 'env.gateway' parameter, but does not allow jump through several servers.

Installation
------------
To install JumpSSH, simply:

.. code:: bash

    $ pip install jumpssh


Examples
--------
establish ssh session with a remote host through a gateway:

.. code:: python

    >>> from jumpssh import SSHSession

    # establish ssh connection between your local machine and the jump server
    >>> gateway_session = SSHSession('gateway.example.com',
    ...                              'my_user', password='my_password').open()

    # from jump server, establish connection with a remote server
    >>> remote_session = gateway_session.get_remote_session('remote.example.com',
    ...                                                     password='my_password2')


run commands on remote host:

.. code:: python

    # command will be executed remotely and output will be returned locally and printed
    >>> print(remote_session.get_cmd_output('ls -lta'))
    total 28
    drwxr-xr-x. 412 root    root    12288 Mar 21 14:25 ..
    drwx------.   2 my_user my_user    28 Mar  6 19:25 .ssh
    drwx------.   3 my_user my_user    70 Mar  6 19:25 .
    -rw-r--r--.   1 my_user my_user    18 Jul 12  2016 .bash_logout
    -rw-r--r--.   1 my_user my_user   193 Jul 12  2016 .bash_profile
    -rw-r--r--.   1 my_user my_user   231 Jul 12  2016 .bashrc

    # get exit code of the remotely executed command (here to check if a package is installed)
    >>> remote_session.get_exit_code('yum list installed package_name')
    0

remote rest api usage:

.. code:: python

    # calling rest api on remote host that is only accessible from the gateway
    >>> from jumpssh import RestSshClient
    >>> rest_client = RestSshClient(gateway_session)

    # syntax is similar to requests library (http://docs.python-requests.org)
    >>> http_response = rest_client.get('http://remote.example.com/helloworld')
    >>> http_response.status_code
    200
    >>> http_response.text
    u'Hello, World!'

remote files operations:

.. code:: python

    # check if remote path exists
    >>> remote_session.exists('/path/to/a/file')
    True

    # copy file from local machine to remote host through gateway
    >>> remote_session.put('/local/path/to/a/file', '/remote/path/to/the/file')

    # create file on remote host from local content
    >>> remote_session.file('/remote/path/to/the/file',
    ...                     content='remote file content', permissions='600')

    # download remote file on local machine from remote host through gateway
    >>> remote_session.get('/remote/path/to/the/file', '/local/path/')


Tests
-----
jumpssh tests require docker, check `docker documentation <https://docs.docker.com>`_ for how to install it
depending on your OS.
it also requires few python packages. To install them, run:

.. code:: bash

    $ pip install -r requirements_dev.txt

To run the test suite, clone the repository and run:

.. code:: bash

    $ pytest -sv tests/

or simply:

.. code:: bash

    $ tox


Contributing
------------

Bug Reports
^^^^^^^^^^^
Bug reports are hugely important! Before you raise one, though,
please check through the `GitHub issues <https://github.com/AmadeusITGroup/JumpSSH/issues>`_,
both open and closed, to confirm that the bug hasn't been reported before.

Feature Requests
^^^^^^^^^^^^^^^^
If you think a feature is missing and could be useful in this module, feel free to raise a feature request through the
`GitHub issues <https://github.com/AmadeusITGroup/JumpSSH/issues>`_

Code Contributions
^^^^^^^^^^^^^^^^^^
When contributing code, please follow `this project-agnostic contribution guide <http://contribution-guide.org/>`_.


CHANGES.rst
1.6.5 (11/03/2020)
------------------
- [Bug] :issue:`152`: Remove pkg_info.json file and replace it with python file to avoid access issue at runtime
- [Improvement] add python 3.9 validation

1.6.4 (08/24/2020)
------------------
- [Bug] :issue:`109`: Fix automated session closure handled by python garbage collection
- [Bug] :issue:`120`: Fix get_remote_session not respecting 'timeout' parameter
- [Bug] :issue:`139`: Fix run_cmd raising AuthenticationException if no agent is running
- [Improvement][Tests]: use flaky package to automatically rerun flaky tests

1.6.3 (03/12/2020)
------------------
- [Improvement]: remove pytest-runner from setup_requires as this is deprecated for security reasons, see https://github.com/pytest-dev/pytest-runner
- [Improvement]: use only fixed test dependencies in requirements_dev.txt

1.6.1 (04/08/2019)
------------------
- [Bug] :issue:`51`: 'get' file was failing if the remote file is binary. Thanks to :user:`pshaobow` for the report.
- [Feature]: Ability to use any parameter of `paramiko.client.SSHClient.connect` in `get_remote_session`, was forgotten during implementation of :issue:`43`.
- [Improvement]: tests migrated to docker-compose to setup docker environment

1.5.1 (01/14/2019)
------------------
- [Feature] :issue:`43`: Ability to use any parameter of paramiko.client.SSHClient.connect in SSHSession.

1.4.1 (03/31/2018)
------------------
- [Bug] :issue:`33`: Fix download of file owned by root with `SSHSession.get`
- [Bug] : Automatically open closed session when calling SSHSession.put. Thanks to :user:`fmaupas` for the fix.

1.4.0 (01/29/2018)
------------------
- [Feature] :issue:`29`: Expose compression support from Paramiko (inherited from SSH).
  Thanks to :user:`fmaupas` for the contribution.

1.3.2 (12/17/2017)
------------------
- [Bug] :issue:`23`: do not print `byte` but `str` in continuous output when running command with python3.
  Thanks to :user:`nicholasbishop` for the report.

1.3.1 (09/15/2017)
------------------
- fix interruption of remote command when transport channel is already closed

1.3.0 (09/14/2017)
------------------
- allow to conceal part of the command run in logs specifying list of pattern in silent parameter (regexp format)
  For example, if a password is specified in command you may want to conceal it in logs but still want to log the
  rest of the command run
- ability to customize success exit code when calling run_cmd so that an exit code different from 0 do not raise
  any exception. Success exit code can be an int or even a list of int if several exit codes are considered a success.
- ability to retry remote command until success or max retry is reached
- ability to forward Ctrl-C to remote host in order to interrupt remote command before stopping local script

1.2.1 (07/27/2017)
------------------
- reduce logging level of some logs
- propagate missing 'silent' parameter in restclient module to run_cmd to control logging 

1.2.0 (07/24/2017)
------------------
- automatically open inactive session when running command on it
- automatically open inactive jump session when requesting remote session

1.1.0 (07/20/2017)
------------------
- Each ssh session can be used as a jump server to access multiple remote sessions in parallel. Only 1 remote
  session per jump server was allowed before.
- ability to customize retry interval when opening a ssh session

1.0.2 (07/14/2017)
------------------
- Fix run of shell builtins commands (source, ...) when impersonating another user as they cannot be executed
  without the shell and by default, sudo do not run shell

1.0.1 (06/11/2017)
------------------
- Fix BadHostKeyException raised by paramiko when reusing same ssh session object to connect to a different
  remote host having same IP than previous host (just TCP port is different)

1.0.0 (05/24/2017)
------------------
- First release


jumpssh/restclient.py
# external import
import os
import json
import logging
import re
try:
    from urllib import quote_plus  # Python2
except ImportError:
    from urllib.parse import quote_plus  # Python3
try:
    from httplib import HTTPResponse as http_HTTPResponse  # Python2
except ImportError:
    from http.client import HTTPResponse as http_HTTPResponse  # Python3
try:
    from StringIO import StringIO as FakeSocketParam  # Python2
except ImportError:
    from io import BytesIO as FakeSocketParam  # Python3

from . import exception, SSHSession

logger = logging.getLogger(__name__)

# list of ANSI escape sequence based on http://ascii-table.com/ansi-escape-sequences-vt-100.php
ANSI_REGEX = r'\x1b(' \
             r'(\[\??\d+[hl])|' \
             r'([=<>a-kzNM78])|' \
             r'([\(\)][a-b0-2])|' \
             r'(\[\d{0,2}[ma-dgkjqi])|' \
             r'(\[\d+;\d+[hfy]?)|' \
             r'(\[;?[hf])|' \
             r'(#[3-68])|' \
             r'([01356]n)|' \
             r'(O[mlnp-z]?)|' \
             r'(/Z)|' \
             r'(\d+)|' \
             r'(\[\?\d;\d0c)|' \
             r'(\d;\dR))'


class RestSshClient(object):
    def __init__(
            self,
            ssh_session=None,
            **kwargs
    ):
        """

        :param ssh_session:
        :param host:
        :param username:
        :param kwargs:
        """
        if ssh_session:
            self.ssh_session = ssh_session
        else:
            self.ssh_session = SSHSession(host=kwargs.pop('host'),
                                          username=kwargs.pop('username'), **kwargs)

        self.host = self.ssh_session.host
        self.username = self.ssh_session.username

    def __enter__(self):
        self.ssh_session.open()
        return self

    def __exit__(self, *args):
        self.ssh_session.close()

    def __repr__(self):
        return '%s(host=%s, username=%s, ssh_session=%s)' \
               % (self.__class__.__name__, self.host, self.username, repr(self.ssh_session))

    def request(self, method, uri, **kwargs):
        """Perform http request and send back http response.

        :param method: http method.
        :param uri: remote URL to target.
        :param params: (optional) Dictionary to be sent in the query string.
        :param data: (optional) Content to send in the body of the http request.
        :param headers: (optional) Dictionary of HTTP Headers to send with the http request.
        :param remote_file: (optional) File on the remote host with content to send in the body of the http request.
        :param local_file: (optional) Local file with content to send in the body of the http request.
        :param document_info_only: (optional) if True, only HTTP Headers are returned in http response (default=False).
        :param auth: (optional) Auth tuple to enable Basic/Digest/Custom HTTP Auth.
        :param verify: (optional) whether the SSL cert will be verified.
        :param silent: if True, does not log the command run (useful if sensitive information are used in command)
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse

        Usage::

          >>> from jumpssh import RestSshClient
          >>> with RestSshClient(host='gateway.example.com', username='my_user') as rest_client:
          >>> ... http_response = rest_client.request('GET', 'http://remote.example.com')
          >>> ... http_response.status_code
          200
        """

        # build curl command
        # force usage of http 1.0 as chunked transfer encoding not yet supported here
        cmd = 'curl -is --http1.0 '

        # disable verification of SSL cert
        if not kwargs.get('verify', True):
            cmd += "-k "

        # only return status code and headers (no body)
        if kwargs.get('document_info_only'):
            cmd += "-I "

        # basic authentication
        if kwargs.get('auth'):
            auth = kwargs.get('auth')
            if len(auth) != 2:
                raise exception.RestClientError("Invalid auth parameter. "
                                                "Tuple with 2 elements (user, password) is expected.")
            cmd += "-u %s:%s " % (auth[0], auth[1])

        # specify http method
        cmd += '-X %s ' % method.upper()

        # build headers list
        if kwargs.get('headers'):
            headers = kwargs.get('headers')
            for key, value in headers.items():
                cmd += '-H "%s:%s" ' % (key, value)

        # add targeted uri
        cmd += '"%s' % uri

        # add uri parameters
        if kwargs.get('params'):
            params = kwargs.get('params')
            if any(params):
                cmd += '?'
                cmd += '&'.join(['%s=%s' % (quote_plus(key), quote_plus(value)) for key, value in params.items()])

        # close quotes around uri with params
        cmd += '" '

        # build body
        if kwargs.get('local_file'):
            local_file = kwargs.get('local_file')
            if not os.path.exists(local_file) or not os.path.isfile(local_file):
                raise exception.RestClientError("Invalid file path given '%s'" % local_file)
            sftp_client = self.ssh_session.get_sftp_client()
            file_name = os.path.basename(local_file)
            sftp_client.put(local_file, file_name)
            cmd += '-d @%s ' % file_name
        elif kwargs.get('remote_file'):
            remote_file = kwargs.get('remote_file')
            if not self.ssh_session.exists(remote_file):
                raise exception.RestClientError("Invalid remote file path given '%s' on host '%s'"
                                                % (remote_file, self.ssh_session.host))
            cmd += '-d @%s ' % remote_file
        elif kwargs.get('data'):
            data = kwargs.get('data')
            cmd += "-d '%s' " % data.replace("'", "\'")

        # execute remote http query and get raw response
        exit_code, output = self.ssh_session.run_cmd(
            cmd,
            # do not raise exception if error code different from 0 as some queries can be successful
            # with other exit codes
            raise_if_error=False,
            # propagate 'silent' parameter to run_cmd
            silent=kwargs.get('silent', False))

        # check exit code has a proper value
        # most of successful commands will return exit code 0
        # except when using HEAD http method as file transfer is shorter or larger than expected
        # curl is returning exit code 18 : CURLE_PARTIAL_FILE (18)
        if exit_code != 0 and not (exit_code == 18 and method.upper() == 'HEAD'):
            raise exception.RestClientError(
                '"Remote command ({command})" returned exit status ({exit_code}): {error}'.format(
                    exit_code=exit_code, command=cmd, error=output)
            )

        # cleanup file copied once rest query has been done
        if kwargs.get('local_file'):
            sftp_client.remove(file_name)

        # return structured http response
        return HTTPResponse(output)

    def get(self, uri, **kwargs):
        r"""Sends a GET request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('get', uri, **kwargs)

    def options(self, uri, **kwargs):
        r"""Sends a OPTIONS request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('options', uri, **kwargs)

    def head(self, uri, **kwargs):
        r"""Sends a HEAD request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('head', uri, **kwargs)

    def post(self, uri, **kwargs):
        r"""Sends a POST request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('post', uri, **kwargs)

    def put(self, uri, **kwargs):
        r"""Sends a PUT request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('put', uri, **kwargs)

    def patch(self, uri, **kwargs):
        r"""Sends a PATCH request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('patch', uri, **kwargs)

    def delete(self, uri, **kwargs):
        r"""Sends a DELETE request.

        :param uri: URL of the http request.
        :param \**kwargs: Optional arguments that :func:`~request` takes.
        :return: :class:`HTTPResponse <HTTPResponse>` object
        :rtype: restclient.HTTPResponse
        """
        return self.request('delete', uri, **kwargs)


class HTTPResponse:
    def __init__(self, http_response_str):
        httplib_HTTPResponse = self.__parse_response(http_response_str)
        self.headers = {}
        self.status_code = httplib_HTTPResponse.status
        self.reason = httplib_HTTPResponse.reason
        # to get full response body from httplib, force read of full payload, and manually override length
        received_length = httplib_HTTPResponse.length
        httplib_HTTPResponse.length = None
        self.text = httplib_HTTPResponse.read().strip().decode('utf-8')
        httplib_HTTPResponse.length = received_length

        for key, value in httplib_HTTPResponse.getheaders():
            self.headers[key] = value

    def __parse_response(self, str_response):
        class FakeSocket(FakeSocketParam):
            def makefile(self, *args, **kw):
                return self

        # remove double '\r' as http.client only searching for the following sequences
        # to identify end of headers list : (b'\r\n', b'\n', b'')
        # so this is failing when receiving for example b'\r\r\n'
        # the change python made to break it in 2.7.13 can be seen here
        # https://fossies.org/diffs/Python/2.7.12_vs_2.7.13/Lib/httplib.py-diff.html
        raw_response = str_response.replace('\r\r\n', '\r\n')

        # make sure any unexpected ansi character added in headers are removed as preventing proper parsing of response
        socket_response = raw_response.split('\n\n', 1)
        if len(socket_response) == 1:
            header = socket_response[0]
            body = ''
        else:
            header, body = socket_response
        ansi_escape = re.compile(ANSI_REGEX, flags=re.IGNORECASE)
        socket_response = ansi_escape.sub('', header) + '\n\n' + body

        socket = FakeSocket(socket_response.encode('utf-8'))
        response = http_HTTPResponse(socket)
        response.begin()
        return response

    def check_for_success(self):
        if self.status_code not in [200, 201]:
            raise exception.RestClientError("http error received: %s" % self)

    def is_valid_json_body(self):
        try:
            json.loads(self.text)
        except ValueError:
            return False
        return True

    def json(self, **kwargs):
        try:
            return json.loads(self.text, **kwargs)
        except ValueError:
            raise exception.RestClientError("http response body is not in a valid json format : %s" + self.text)

    def __str__(self):
        result = "%s %s\n" % (self.status_code, self.reason)
        try:
            result += json.dumps(json.loads(self.text), indent=4, sort_keys=True)
        except ValueError:
            result += self.text
        return result
